### ISSUE TITLE
Nested parallel processes: how to handle them?

### ISSUE THREAD

**Issue by gioxc88** [dask/dask#6705]
I couldn't find any satisfactory question (and answer) either in the opened issues or on stackoverflow, so I thought I might as well open an issue myself.

I'll illustrate the problem with my use-case (which is oversimplified below).

1. My dataset contains 30 observations which I divide in 3 subsamples of 10 observation each.
(the real dataset is much bigger and the number of subsamples is much higher)

```
>>> print(len(X))
>>> 30

     subsample1            subsample2            subsample3 
x x x x x x x x x x | x x x x x x x x x x | x x x x x x x x x x | 
0 1 2 3 4 5 6 7 8 9 | 0 1 2 3 4 5 6 7 8 9 | 0 1 2 3 4 5 6 7 8 9 | 

```

2. For each subsample I train a model which needs to be tuned via cross-validation, for example using `GridSearchCV` or `HyperbandSearchCV`. 
Say that I want to validate the `l1` penalty of a `Lasso` regressor using a grid of 50 points, such that I train 50 models for every subsample.

```
from dask import compute, delayed
from dask_ml.model_selection import GridSearchCV
from sklearn.linear_model import Lasso

reg = Lasso()
param_space = np.logspace(1, -9, num=50)
param_grid ={'alpha': param_space}
model = GridSearchCV(reg)
```

In this case the schema would look something like:

```          
              | model0
              | model1
subsample3 -> | ...
              | model49
------------------------
              | model0
              | model1
subsample3 -> | ...
              | model49
------------------------
              | model0
              | model1
subsample3 -> | ...
              | model49
```

As I said, in the real world application I have more than 100 subsamples on which I cross validate the model by training 50 different hyperparameter combinations for each subsample.

I could easily write a code like this:

```
from dask import compute, delayed
from . import load_data, get_subsample

X, y = load_dada()  # load_data is an imaginary function for loading the data
n_subsample = 100

results = []
for subsample in range(n_subsamples):
    X_train, y_train = get_subsample(X, y, subsample)  # get_subsample is an imaginary function which slices X and y
    results.append(delayed(model.fit)(X_train, y_train))
results = compute(*results)
```

To summarize:
1. Since I parallelized computations across subsamples `dask` is spawning one process for each subsample. 
2. Since within each subsample I use `GridSearchCV`, which also makes use of multiprocessing, this will also try to spawn processes within the process.

In this case we could talk about nested processes.
My questions are:
1. How does dask handle this? If it doesn't handle it already, should it be able to?
2. Is this a bad design and does not have anything to do with dask in the first place?

Many thanks

**Comment by mrocklin** [comment]
I recommend using the dask.distributed scheduler for this.  It will route
dask calls back to the main scheduler when they are run, so everything will
share resources.

**Comment by gioxc88** [comment]
> I recommend using the dask.distributed scheduler for this. It will route dask calls back to the main scheduler when they are run, so everything will share resources.

Thanks so much for taking the time for looking into this.
I tried the approach you suggested but unfortunately it does not work.

For testing  I used `dask_ml.model_selection.GridSearchCV` as described above
Before that I started a local cluster by doing 
```
from dask.distributed import Client
client = Client()
```

What happens is that the graph is computed but then the tasks never starts, they hang indefinitely.
Below the screen of what I see in the dashboard. 

![image](https://user-images.githubusercontent.com/41852601/95522161-41ac0f80-09c3-11eb-86b7-871903da22b6.png)

Do you think I can provide any additional information to help identify what causes the problem?

Many thanks
Gio

**Comment by mrocklin** [comment]
Hrm, that's odd.  To aid figuring out what's going on are you able to provide a minimal reproducer?  That would let maintainers run into the same problem that you're running into and more easily be able to see what is going on without guessing.  You might want to use some of the sklearn datasets to assist here.

**Comment by gioxc88** [comment]
> Hrm, that's odd. To aid figuring out what's going on are you able to provide a minimal reproducer? That would let maintainers run into the same problem that you're running into and more easily be able to see what is going on without guessing. You might want to use some of the sklearn datasets to assist here.

I am very happy to provide a reproducible example.
I just tested this on  JupyterLab with:
python 3.8.6
dask 2.29.0
dask-ml 1.7.0
(even though I got the same problem with previous versions)

```
import numpy as np
import pandas as pd

from dask import compute, delayed
from dask.distributed import Client
from dask_ml.model_selection import GridSearchCV
from sklearn.linear_model import Lasso

# random number generator
rng = np.random.default_rng(7)

# client
client = Client()
client

# data
n_obs = 1000
n_features = 10
X = pd.DataFrame(rng.normal(size=(n_obs, n_features)))
y = pd.Series(rng.normal(size=(n_obs, )))

# estimators 
estimator = Lasso()
param_grid = {'alpha': np.logspace(1, -5, num=10)}
search = GridSearchCV(estimator, param_grid=param_grid)

# subsamples 
n_subsamples = 10
subsample_size = n_obs // n_subsamples
subsamples_starts = range(0, n_obs, subsample_size)

# parallel computation
results = []
for subsample_start in subsamples_starts:
    X_train = X[subsample_start:subsample_start+subsample_size]
    y_train = y[subsample_start:subsample_start+subsample_size]
    results.append(delayed(search.fit)(X_train, y_train))
results = compute(*results)
```
And this is what is shown on the dashboard. It's stuck and I need to interrupt the kernel as the cell never gets executed. The same behavior is present if I use PyCharm (just to say I don't think it is related to JupyterLab)

![image](https://user-images.githubusercontent.com/41852601/95788946-d45af000-0cdc-11eb-9d59-c935060dca6e.png)

Thank you very much to anyone who will look into this.
Gio

**Comment by mrocklin** [comment]
Hrm, on a first attempt this unfortunately ran fine for me.

```python
In [1]: import numpy as np
   ...: import pandas as pd
   ...: 
   ...: from dask import compute, delayed
   ...: from dask.distributed import Client
   ...: from dask_ml.model_selection import GridSearchCV
   ...: from sklearn.linear_model import Lasso
   ...: 
   ...: # random number generator
   ...: rng = np.random.default_rng(7)
   ...: 
   ...: # client
   ...: client = Client()
   ...: client
   ...: 
   ...: # data
   ...: n_obs = 1000
   ...: n_features = 10
   ...: X = pd.DataFrame(rng.normal(size=(n_obs, n_features)))
   ...: y = pd.Series(rng.normal(size=(n_obs, )))
   ...: 
   ...: # estimators 
   ...: estimator = Lasso()
   ...: param_grid = {'alpha': np.logspace(1, -5, num=10)}
   ...: search = GridSearchCV(estimator, param_grid=param_grid)
   ...: 
   ...: # subsamples 
   ...: n_subsamples = 10
   ...: subsample_size = n_obs // n_subsamples
   ...: subsamples_starts = range(0, n_obs, subsample_size)
   ...: 
   ...: # parallel computation
   ...: results = []
   ...: for subsample_start in subsamples_starts:
   ...:     X_train = X[subsample_start:subsample_start+subsample_size]
   ...:     y_train = y[subsample_start:subsample_start+subsample_size]
   ...:     results.append(delayed(search.fit)(X_train, y_train))
   ...: results = compute(*results)
   ...: 

In [2]: results
Out[2]: 
(GridSearchCV(estimator=Lasso(),
              param_grid={'alpha': array([1.00000000e+01, 2.15443469e+00, 4.64158883e-01, 1.00000000e-01,
        2.15443469e-02, 4.64158883e-03, 1.00000000e-03, 2.15443469e-04,
        4.64158883e-05, 1.00000000e-05])}),
 GridSearchCV(estimator=Lasso(),
              param_grid={'alpha': array([1.00000000e+01, 2.15443469e+00, 4.64158883e-01, 1.00000000e-01,
        2.15443469e-02, 4.64158883e-03, 1.00000000e-03, 2.15443469e-04,
        4.64158883e-05, 1.00000000e-05])}),
 GridSearchCV(estimator=Lasso(),
              param_grid={'alpha': array([1.00000000e+01, 2.15443469e+00, 4.64158883e-01, 1.00000000e-01,
        2.15443469e-02, 4.64158883e-03, 1.00000000e-03, 2.15443469e-04,
        4.64158883e-05, 1.00000000e-05])}),
 GridSearchCV(estimator=Lasso(),
              param_grid={'alpha': array([1.00000000e+01, 2.15443469e+00, 4.64158883e-01, 1.00000000e-01,
        2.15443469e-02, 4.64158883e-03, 1.00000000e-03, 2.15443469e-04,
        4.64158883e-05, 1.00000000e-05])}),
 GridSearchCV(estimator=Lasso(),
              param_grid={'alpha': array([1.00000000e+01, 2.15443469e+00, 4.64158883e-01, 1.00000000e-01,
        2.15443469e-02, 4.64158883e-03, 1.00000000e-03, 2.15443469e-04,
        4.64158883e-05, 1.00000000e-05])}),
 GridSearchCV(estimator=Lasso(),
              param_grid={'alpha': array([1.00000000e+01, 2.15443469e+00, 4.64158883e-01, 1.00000000e-01,
        2.15443469e-02, 4.64158883e-03, 1.00000000e-03, 2.15443469e-04,
        4.64158883e-05, 1.00000000e-05])}),
 GridSearchCV(estimator=Lasso(),
              param_grid={'alpha': array([1.00000000e+01, 2.15443469e+00, 4.64158883e-01, 1.00000000e-01,
        2.15443469e-02, 4.64158883e-03, 1.00000000e-03, 2.15443469e-04,
        4.64158883e-05, 1.00000000e-05])}),
 GridSearchCV(estimator=Lasso(),
              param_grid={'alpha': array([1.00000000e+01, 2.15443469e+00, 4.64158883e-01, 1.00000000e-01,
        2.15443469e-02, 4.64158883e-03, 1.00000000e-03, 2.15443469e-04,
        4.64158883e-05, 1.00000000e-05])}),
 GridSearchCV(estimator=Lasso(),
              param_grid={'alpha': array([1.00000000e+01, 2.15443469e+00, 4.64158883e-01, 1.00000000e-01,
        2.15443469e-02, 4.64158883e-03, 1.00000000e-03, 2.15443469e-04,
        4.64158883e-05, 1.00000000e-05])}),
 GridSearchCV(estimator=Lasso(),
              param_grid={'alpha': array([1.00000000e+01, 2.15443469e+00, 4.64158883e-01, 1.00000000e-01,
        2.15443469e-02, 4.64158883e-03, 1.00000000e-03, 2.15443469e-04,
        4.64158883e-05, 1.00000000e-05])}))
```

**Comment by gioxc88** [comment]
That is super strange.
Both on my workstation and on my notebook the behavior is the same when I execute the code above. Maybe the problem is python 3.8?

Is there any other thing I can do to help you identify the problem?
Many thanks 

Gio

**Comment by TomAugspurger** [comment]
I also saw he computation hanging. A couple things stand out:

1. You're delaying `dask_ml.model_selection.GridSearchCV.fit`. It's strange to wrap task-generating things like dask-ml's GridSearchCV inside a `delayed` object. Does just using `sklearn.model_selection.GridSearchCV` work?
2. Scikit-Learn's `.fit` API relies on mutation of the object to propagate return values. So `delayed(search.fit)` is delaying a reference to an object `search` that's mutated in each fit call. I'd recommend delaying `fit` on a cloned version

```python
# parallel computation
results = []
for subsample_start in subsamples_starts:
    X_train = X[subsample_start:subsample_start+subsample_size]
    y_train = y[subsample_start:subsample_start+subsample_size]
    results.append(delayed(sklearn.base.clone(search).fit)(X_train, y_train))
results = compute(*results)
```

That completes for me.

**Comment by gioxc88** [comment]
> I also saw he computation hanging. A couple things stand out:
> 
> 1. You're delaying `dask_ml.model_selection.GridSearchCV.fit`. It's strange to wrap task-generating things like dask-ml's GridSearchCV inside a `delayed` object. Does just using `sklearn.model_selection.GridSearchCV` work?
> 2. Scikit-Learn's `.fit` API relies on mutation of the object to propagate return values. So `delayed(search.fit)` is delaying a reference to an object `search` that's mutated in each fit call. I'd recommend delaying `fit` on a cloned version

Dear @TomAugspurger, thank you very much for looking into this. I am aware of both points and I have an answer for all of them.

I'll address 2 first: very simply, for the sake of keeping the example minimal I did not clone the estimator here but in the original code it is deepcopied, but thanks for bringing that up.

1. The only difference in using `dask GridSearchCV` instead of `sklearn GridSearchCV` is that in the first case the process never starts (it hangs) in the second case it starts with a warning though:

`loky-backed parallel loops cannot be called in a multiprocessing setting n_jobs=1`

this means that effectively multiporcessing is not used even if I use `sklearn GridSearchCV`. Furthermore, I would prefer using  `dask GridSearchCV` because on the dashboard I got much more details about what task is being executed. 

The reason why I want to do parallel computation both on the subsamples and on the parameter space is that in many case I could have a lot of subsamples but a parameter space which is not very broad so GridSeachCV or any other SearchCV instances might end up not using 100% of cpu for a single subsample.

Gio

**Comment by gioxc88** [comment]
any news on this?
Many thanks

Gio

**Comment by gioxc88** [comment]
> I'm not sure, but the strangest thing I see is the `delayed(search).fit(X, y)`. Typically you don't delay things that are already using dask.
> 
> If you're interested, you could try to get the raw dask graph out of https://github.com/dask/dask-ml/blob/master/dask_ml/model_selection/_search.py#L1230 to avoid the delay. And then do one big compute at the end on all of those.

Thank you for the answer
If you look carefully I am delaying `search.fit` and not `search` as your code implies.

After running additional tests I think I found a hint on what could be the cause of the problem.
Basically today I tested the code on linux and if works fine, whereas on Windows it hangs as described above.

Since @mrocklin you tested the code and it worked fine for you could you please confirm that you actually tested it on a linux machine?

also @TomAugspurger you said it was hanging for you, could you please confirm that you tested it on windows?

If that's the case I am afraid my knowledge is too limited to open on PR which could fix the problem on windows and I would ask help from the community to fix it.

Many thanks

**Comment by mrocklin** [comment]
> Since @mrocklin you tested the code and it worked fine for you could you please confirm that you actually tested it on a linux machine?

Yes, I run Ubuntu 20.04 on a Lenovo Thinkpad.

**Comment by gioxc88** [comment]
any idea of why it hangs on windows?

**Comment by mrocklin** [comment]
I do not right now have such an idea.  It is unlikely that I personally
will test this out immediately.  I encourage others to though.

**Comment by gioxc88** [comment]
@mrocklin I doing tests using nested calls to delayed and compute and I also came across the problem of [work stealing](http://distributed.dask.org/en/latest/work-stealing.html).
This results in errors and the overall job fails with a traceback because some process has been closed. And I think is related to work stealing.

I wonder though can this be because I also call compute on multiple times so the scheduler is not able to make the whole computation graph in advance?

Would something returning the futures object for each nested function and then calling compute once solve the problem?

I was wondering if this nested problem can be attacked by using async def or using coroutines.

This would change my implementation by a lot so I wanted your opinion on this before diving into deep waters 

Thanks
Gio
